### PR TITLE
Support SSL client auth for elastic search

### PR DESCRIPTION
**Pull request by schenksj**
**Clearly describe the problem and solution**
This PR adds a capability to interface all Moloch components with an Elasticsearch cluster that leverages https client auth.

`db.pl` requires installing `perl-LWP-Protocol-https` and `perl-Crypt-SSLeay` if SSL is in use.

**Relevant issue number(s) if applicable**
https://github.com/aol/moloch/issues/1089

**Did you run `npm run lint` from the viewer or parliament directory (whichever you are making changes to) and correct any errors**
will not execute on my mac OSX installation

**Did you Ensure that all tests still pass by navigating to the `tests` directory and running `./tests.pl --viewer`**
yes

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: capture/http.c [hunk 1]
```diff
@@ -201,6 +201,15 @@ unsigned char *moloch_http_send_sync(void *serverV, const char *method, const ch
         curl_easy_setopt(easy, CURLOPT_SSL_VERIFYHOST, 0L);
     }
 
+    // Send client certs if so configured
```
Comment (awick) [inline capture/http.c]: The http code is used by other things then ES, so you'll need to add a moloch_http_set_client_cert (or whatever name) call kind of like moloch_http_set_retries and save it into the molochhttpserver_t and use that here.  Then in db.c you'll need to call it for the esServer

File: viewer/db.js [hunk 2]
```diff
@@ -573,18 +583,12 @@ exports.numberOfUsers = function(cb) {
 
 exports.deleteUser = function (name, cb) {
   delete internals.usersCache[name];
-  return internals.usersElasticSearchClient.delete({index: internals.usersPrefix + 'users', type: 'user', id: name, refresh: true}, (err) => {
-    delete internals.usersCache[name]; // Delete again after db says its done refreshing
-    cb(err);
-  });
+  return internals.usersElasticSearchClient.delete({index: internals.usersPrefix + 'users', type: 'user', id: name, refresh: true}, cb);
```
Comment (awick) [inline viewer/db.js]: i think you had a merge error here, its missing the new code

File: viewer/db.js [hunk 3]
```diff
@@ -573,18 +583,12 @@ exports.numberOfUsers = function(cb) {
...
 };
 
 exports.setUser = function(name, doc, cb) {
   delete internals.usersCache[name];
-  return internals.usersElasticSearchClient.index({index: internals.usersPrefix + 'users', type: 'user', body: doc, id: name, refresh: true}, (err) => {
-    delete internals.usersCache[name]; // Delete again after db says its done refreshing
-    cb(err);
-  });
+  return internals.usersElasticSearchClient.index({index: internals.usersPrefix + 'users', type: 'user', body: doc, id: name, refresh: true}, cb);
```
Comment (awick) [inline viewer/db.js]: missing new code

File: viewer/db.js (hunk 2 above)
Comment (schenksj) [reply viewer/db.js]: sorry, I thought I fixed all those... will get on it

File: viewer/db.js (hunk 3 above)
Comment (schenksj) [reply viewer/db.js]: sorry, I thought I fixed all those... will get on it

File: capture/http.c (hunk 1 above)
Comment (schenksj) [reply capture/http.c]: Working on it... will rename the config variables to reflect the being ES only

Comment (schenksj) [reply capture/http.c]: Done. Please let me know if this is what you had in mind.

**Comment by schenksj** [comment]
@awick Whenever you have a chance, I think this is ready for another round of reviews.

File: db/db.pl [hunk 4]
```diff
@@ -2067,6 +2070,14 @@ sub parseArgs {
         $PREFIX .= "_" if ($PREFIX !~ /_$/);
     } elsif ($ARGV[0] =~ /-n$/) {
         $NOCHANGES = 1;
+    } elsif ($ARGV[0] =~ /--insecure$/) {
+        $SECURE = 0;
```
Comment (awick) [inline db/db.pl]: You need to define all these vars up top, the tests are failing

https://cd.screwdriver.cd/pipelines/354/builds/132281/steps/test-viewer

File: db/db.pl [hunk 5]
```diff
@@ -2089,6 +2100,14 @@ sub parseArgs {
...
 
 $main::userAgent = LWP::UserAgent->new(timeout => $ESTIMEOUT + 5, keep_alive => 5);
+if ($CLIENTCERT ne "") {
+    $main::userAgent->ssl_opts(
+        SSL_verify_mode => $SECURE,
+        verify_hostname=> $SECURE,
+        SSL_cert_file => $CLIENTCERT,
+        SSL_key_file => $CLIENTKEY
+    )
+}
```
Comment (awick) [inline db/db.pl]: maybe add an else here that if $SECURE is set to 0 but CLIENTCERT isn't set it still turns off verify?

File: db/db.pl (hunk 4 above)
Comment (schenksj) [reply db/db.pl]: ugh... these merge errors are killing me

File: db/db.pl (hunk 5 above)
Comment (schenksj) [reply db/db.pl]: yes, makes sense

**Comment by schenksj** [comment]
@awick I have made the requested changes (sorry for the merge issue).  I'm getting a failure on something related to API stuff that I don't think I touched...  Any tips on what to do to address?

**Comment by awick** [comment]
That failure is a known race condition that we are still working on.    Thanks for the work!

**Pull request merged**